### ISSUE TITLE
Change Swift Concurrency Availability to macOS 10.15, iOS 13, tvOS 13, and watchOS 6

### ISSUE THREAD

**Issue by PSchmiedmayer** [apple/swift-nio#2000]
### Expected behavior
With the support for Swift Concurrency coming to macOS 10.15, iOS 13, tvOS 13, and watchOS 6 with Xcode 13.2, the Swift Concurrency features should be available when compiling for these targets using Xcode 13.2.

### Actual behavior
The Swift Concurrency features are marked with `@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)` and can therefore not be used on macOS 10.15, iOS 13, tvOS 13, and watchOS 6.

### Steps to reproduce

1. Change the source code of the NIOAsyncAwaitDemo on Xcode 13.2 and change the `@available` annotations to macOS 10.15, iOS 13, tvOS 13, and watchOS 6.
2. The Swift compiler throws compiler errors due to the `@available` in the Swift NIO targets.

### Solution
A simple solution would be to change the `@available` annotations to `@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)` as demonstrated at https://github.com/PSchmiedmayer/swift-nio.
Unfortunately, this breaks builds on the latest stable Xcode version (13.1) due to compiler errors disallowing Swift Concurrency features previous to macOS 12, which was the state when Xcode 13.1 was released.
This behavior is demonstrated by the GitHub Actions that I have added to https://github.com/PSchmiedmayer/swift-nio: https://github.com/PSchmiedmayer/swift-nio/actions/runs/1532925179 (latest == Xcode 13.2, latest_stable == Xcode 13.1)

I am not sure how this could be easily solved. I would be happy to receive any suggestions to successfully create a PR to change the Swift Concurrency availability to macOS 10.15, iOS 13, tvOS 13, and watchOS 6 with Xcode 13.2.

### SwiftNIO version/commit hash

[d08e8ff938bcfa339f47677716d21f531fa493c3](https://github.com/apple/swift-nio/commit/d08e8ff938bcfa339f47677716d21f531fa493c3)

### Swift & OS version (output of `swift --version && uname -a`)

swift-driver version: 1.26.21 Apple Swift version 5.5.2 (swiftlang-1300.0.47.2 clang-1300.0.29.30)
Target: x86_64-apple-macosx12.0
Darwin 21.1.0 Darwin Kernel Version 21.1.0; root:xnu-8019.41.5~1/RELEASE_X86_64 x86_64

### PS

GitHub Issue/PR #2000 🥳🚀

**Comment by Lukasa** [comment]
Thanks for this issue!

Currently there's no easy solution: anyone running the SDKs available in older Xcodes will find their compiles break when we make this change. It's probably acceptable to do so, but we definitely can't do it _now_: we're going to have to wait a little while to reduce the blast radius for our community. I wouldn't expect us to do this until at least a bit of a while after Xcode 13.2 ships, to allow users to move forward appropriately.

**Comment by PSchmiedmayer** [comment]
@Lukasa Thanks for the insights and good to hear that I didn't miss an obvious solution. Makes sense to wait a bit after Xcode 13.2 is released.

**Comment by weissi** [comment]
With this, I think the product should be renamed NIO2000

**Comment by weissi** [comment]
@Lukasa in theory we could guard all the code (with the new, lowered availabilities) behind `#if compiler(>=5.5.1)` or whatever the right version for that Xcode is.

And then, we could add one more file `AysncAwaitSupport+OldXcodes.swift` or so where we put a whole copy of everything with `#if os(macOS) && compiler(>=5.5) && !compiler>=5.5.1)` with the old, macOS 12 availability. Yeah, this sucks but I think it's a one-off effort to copy everything. New additions only need to be made to the new files (with the lower availability).

**Comment by Lukasa** [comment]
Do we honestly think it's worth it?

**Comment by weissi** [comment]
I don't really know how important this is to the community, just commented to note that I think there could be a (awkward) solution if this turns out to be important... It doesn't matter to _me personally_ to have the macOS12 etc availability.

**Comment by PSchmiedmayer** [comment]
@weissi @Lukasa Thanks for your input on this issue! I have tried @weissi idea and created a PR that reflects this approach: #2004.

I agree with @Lukasa that this introduces code duplication. Still, I would argue that this is only temporary and would allow developers to immediately use Swift Concurrency starting with Swift 5.2.2 and Xcode 13.2. We can remove the duplicated code once we assume a wide enough adoption of Xcode 13.2. This approach sounds like a good compromise to me.

I have verified that it works with Xcode 13.1 and Xcode 13.2 in the following GitHub Action run: https://github.com/PSchmiedmayer/swift-nio/actions/runs/1535425293. I removed the GitHub Action when I created the PR as this repo does not use GitHub Actions at the moment.
